### PR TITLE
Do not create close listener handlers for every new request

### PR DESCRIPTION
**Pull request by spericas**
Do not create listener handlers for every new request resulting in a memory leak. Use using channelInactive() instead. Issue #3850.

Signed-off-by: Santiago Pericasgeertsen <santiago.pericasgeertsen@oracle.com>

### CODE REVIEW

**Comment by lambdaupb** [comment]
Wouldn't that have the same issue with connections that have a steady rate of requests, like polling a service once per second?

**Comment by spericas** [comment]
> Wouldn't that have the same issue with connections that have a steady rate of requests, like polling a service once per second?

Despite its name, that method is called only when a channel is closed.

**Comment by lambdaupb** [comment]
> > Wouldn't that have the same issue with connections that have a steady rate of requests, like polling a service once per second?
> 
> Despite its name, that method is called only when a channel is closed.

Exactly.

if I make one million (and more) request at a rate of 1/s the there will be 1mio listeners (and growing).

or is this a per-request sub channel?

**Comment by spericas** [comment]
> > > Wouldn't that have the same issue with connections that have a steady rate of requests, like polling a service once per second?
> > 
> > 
> > Despite its name, that method is called only when a channel is closed.
> 
> Exactly.
> 
> if I make one million (and more) request at a rate of 1/s the there will be 1mio listeners (and growing).
> 
> or is this a per-request sub channel?

Not sure I follow. The listeners that were there before have been removed by this PR; the close event is now handled by the same class as the other connection events.

**Comment by lambdaupb** [comment]
A connection with keep-alive is long lived and the potential requests processed over it is practically unbounded.

Adding a close listener to the long lived connection without removing it upon request completion will lead to unbounded listeners being registered.

The problem wasn't that listeners kept around after a connection was closed (after the keep-alive timeout they went away), but that listeners of long completed futures stayed registered, keeping the corresponding futures from being GCed.

As far as I can tell, I'm not well versed in the async intricacies.

**Comment by spericas** [comment]
> A connection with keep-alive is long lived and the potential requests processed over it is practically unbounded.
> 
> Adding a close listener to the long lived connection without removing it upon request completion will lead to unbounded listeners being registered.

 This is correct. This is why this PR *does not* add those listeners anymore. That code is being removed by this PR.
 
> The problem wasn't that listeners kept around after a connection was closed (after the keep-alive timeout they went away), but that listeners of long completed futures stayed registered, keeping the corresponding futures from being GCed.

 That's right. Again, no more listeners now.

**Comment by lambdaupb** [comment]
Aaah, I misunderstood then. Sorry for the noise.

Thanks for the quick fix.

File: webclient/webclient/src/main/java/io/helidon/webclient/NettyClientHandler.java [hunk 1]
```diff
@@ -236,6 +226,17 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpObject msg) throws IO
...
 
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+        // Connection closed without last HTTP content received. Some server problem
+        // so we need to fail the publisher and report an exception.
+        if (!responseCloser.isClosed()) {
+            WebClientException exception = new WebClientException("Connection reset by the host");
+            publisher.fail(exception);
+        }
```
Comment (danielkec) [inline webclient/webclient/src/main/java/io/helidon/webclient/NettyClientHandler.java:237]: There is null check in the `channelReadComplete` I would expect same is needed here, previous version captured it from the `channelRead0` context.
```suggestion
        if (publisher != null && !responseCloser.isClosed()) {
            WebClientException exception = new WebClientException("Connection reset by the host");
            publisher.fail(exception);
        }
```

**Review by danielkec — APPROVED** [review]

**Pull request merged**